File: src/pubsub.erl
-module(pubsub).
-behavior(gen_event).

%% public API:
-export([start/0, stop/1, publish/2, 
        add_sub/2, add_sub/1, 
        remove_sub/1]).

%% gen_event callbacks:
-export([init/1, handle_event/2, handle_call/2, handle_info/2, 
         terminate/2, code_change/3]).

%% Simple PubSub mechanism.

%% API

%% Starts the event manager (not included in a supervision tree).
start() -> 
    {ok, Manager} = gen_event:start(),
    Args = ok,
    gen_event:add_handler(Manager, ?MODULE, Args),
    {ok, Manager}.

%% Publishes a message to all subscribers.
publish(EventManager, Message) when is_pid(EventManager) ->
    gen_event:sync_notify(EventManager, {publish, Message}).

%% Adds a function that can process an incoming message as a subscriber to the 
%% pubsub mechanism. This can be used to e.g. better isolate your code in a 
%% single module. The subscribe function itself should do as little work as 
%% possible (for example forwarding a message).
%%
%% Example:
%% 
%% ... 
%% SelfPid = self(),
%% pubsub:add_sub(EvtMgr, fun(Msg) -> 
%%      gen_server:call(SelfPid, {new_msg, Msg}) 
%% end).
%% ...
%% 
%% handle_call({new_msg, Msg}, From, State) -> ...
%%
add_sub(EventManager, SubscribeFunction) 
        when is_pid(EventManager) 
        andalso is_function(SubscribeFunction, 1) ->
    gen_event:sync_notify(EventManager, {add_sub, self(), SubscribeFunction}).

%% Adds the calling process as a subscriber to the pubsub mechanism.
%% Incoming messages are delivered as is (using the ! operator).
add_sub(EventManager) -> 
    Pid = self(),
    F = fun(Message) ->
        Pid ! Message
    end,
    add_sub(EventManager, F).

%% Removes the calling process from the pubsub mechanism, should be the same
%% process as the one that subscribed to the pubsub process.
remove_sub(EventManager) when is_pid(EventManager)  -> 
    gen_event:sync_notify(EventManager, {remove_sub, self()}).

%% Stops the event manager.
stop(EventManager) when is_pid(EventManager) -> gen_event:stop(EventManager).

%% gen_event callbacks:

init(ok) -> {ok, #{}}. % Begin state = empty map.

handle_event({publish, Message}, SubscriberMap) ->
    % TODO check next line performance, 
    % or use maps:map if too slow
    [Func(Message) || Func <- maps:values(SubscriberMap)],
    {ok, SubscriberMap};
handle_event({add_sub, PidKey, Func}, Subscribers) ->
    {ok, maps:put(PidKey, Func, Subscribers)};

handle_event({remove_sub, PidKey}, Subscribers) ->
    {ok, maps:remove(PidKey, Subscribers)}.

handle_call(_Request, _State) -> ok. % not used.
handle_info(_Request, _State) -> ok. % not used.

terminate(_Arg, _State) -> ok.
code_change(_OldVersion, State, _Extra) -> {ok, State}.

File: lib/pubsub.ex
defmodule Snake.Pubsub do
  use GenEvent

  @moduledoc """
  Simple publish subscribe mechanism.
  """

  @doc """
  Starts the event manager for the pubsub mechanism (can be outside of a
  supervision tree.
  """
  def start_link do
    {:ok, manager} = GenEvent.start_link
    args = :ok
    manager |> GenEvent.add_handler __MODULE__, args
    {:ok, manager}
  end

  @doc """
  Publishes a message to all subscribers of the pubsub event manager.
  """
  def publish(manager, msg) when is_pid(manager) do
    manager |> GenEvent.sync_notify {:publish, msg}
  end

  @doc """
  Adds a function that can process an incoming message as a subscriber to the 
  pubsub mechanism. This can be used to e.g. better isolate your code in a 
  single module. The subscribe function itself should do as little work as 
  possible (for example forwarding a message).

  Example:
 
  ... 
  pid = self()
  manager |> Pubsub.add_sub fun(msg) -> 
    :ok = pid |> GenServer.call {new_msg, msg} 
  end
  ...
 
  def handle_call({new_msg, msg}, from, state) -> ...
  """
  def add_sub(manager, pid, sub_function)
      when is_pid(manager)
      and is_pid(pid)
      and is_function(sub_function, 1) do
    manager |> GenEvent.sync_notify {:add_sub, pid, sub_function}
  end

  @doc """
  Adds the calling process as a subscriber to the pubsub mechanism.
  Incoming messages are delivered as is (using send operator).
  """
  def add_sub(manager, pid) do
    manager |> add_sub fn(msg) ->
      pid |> send msg
    end
  end

  @doc """
  Removes a process from the pubsub mechanism
  """
  def remove_sub(manager, pid) when is_pid(manager) do
    manager |> GenEvent.sync_notify {:remove_sub, pid}
  end

  @doc """
  Stops the pubsub manager process.
  """
  def stop(manager) when is_pid(manager), do: manager |> GenEvent.stop

  # GenEvent callbacks

  @doc false
  def init(:ok), do: {:ok, HashDict.new} # Begin state = empty hashdict.

  @doc false
  def handle_event({:publish, msg}, subscribers = %HashDict{}) do 
    subscribers |> Enum.map fn {_pid_key, function} ->
      function |> apply [msg]
    end
    {:ok, subscribers}
  end
  def handle_event({:add_sub, pid_key, func}, subscribers = %HashDict{}) do
    {:ok, subscribers |> HashDict.put(pid_key, func)}
  end
  def handle_event({:remove_sub, pid_key}, subscribers = %HashDict{}) do
    {:ok, subscribers |> HashDict.delete pid_key}
  end
end
